storage/fcds/fcds.go

```diff
@@ -0,0 +1,362 @@
+		offset, err = sh.f.Seek(0, io.SeekEnd)
+	} else {
+		// seek to the offset position
+		// to replace the chunk data at that position
+		_, err = sh.f.Seek(offset, io.SeekStart)
+	}
+	if err != nil {
+		return err
+	}
+
+	if _, err = sh.f.Write(section); err != nil {
+		return err
+	}
+	if reclaimed && s.freeCache != nil {
+		s.freeCache.remove(shard, offset)
+	}
+	return s.meta.Set(addr, shard, reclaimed, &Meta{
+		Size:   uint16(size),
+		Offset: offset,
+	})
+}
+
+// getOffset returns an offset where chunk data can be written to
+// and a flag if the offset is reclaimed from a previously removed chunk.
+// If offset is less then 0, no free offsets are available.
+func (s *Store) getOffset(shard uint8) (offset int64, reclaimed bool, err error) {
+	if !s.shardHasFreeOffsets(shard) {
+		return -1, false, nil
+	}
+
+	offset = -1
+	if s.freeCache != nil {
+		offset = s.freeCache.get(shard)
+	}
+
+	if offset < 0 {
+		offset, err = s.meta.FreeOffset(shard)
+		if err != nil {
+			return 0, false, err
+		}
+	}
+	if offset < 0 {
+		s.markShardWithFreeOffsets(shard, false)
+		return -1, false, nil
+	}
+
+	return offset, true, nil
+}
+
+// Delete makes the chunk unavailable.
+func (s *Store) Delete(addr chunk.Address) (err error) {
+	if err := s.protect(); err != nil {
+		return err
+	}
+	defer s.unprotect()
+
+	shard := getShard(addr)
+	s.markShardWithFreeOffsets(shard, true)
+
+	if s.freeCache != nil {
+		m, err := s.getMeta(addr)
+		if err != nil {
+			return err
+		}
+		s.freeCache.set(shard, m.Offset)
+	}
+	return s.meta.Remove(addr, shard)
+}
+
+// Count returns a number of stored chunks.
+func (s *Store) Count() (count int, err error) {
+	return s.meta.Count()
+}
+
+// Iterate iterates over stored chunks in no particular order.
+func (s *Store) Iterate(fn func(chunk.Chunk) (stop bool, err error)) (err error) {
+	if err := s.protect(); err != nil {
+		return err
+	}
+	defer s.unprotect()
+
+	for _, sh := range s.shards {
+		sh.mu.Lock()
+	}
+	defer func() {
+		for _, sh := range s.shards {
+			sh.mu.Unlock()
+		}
+	}()
+
+	return s.meta.Iterate(func(addr chunk.Address, m *Meta) (stop bool, err error) {
+		data := make([]byte, m.Size)
+		_, err = s.shards[getShard(addr)].f.ReadAt(data, m.Offset)
+		if err != nil {
+			return true, err
+		}
+		return fn(chunk.NewChunk(addr, data))
+	})
+}
+
+// Close disables of further operations on the Store.
+// Every call to its methods will return ErrStoreClosed error.
+// Close will wait for all running operations to finish before
+// closing its MetaStore and returning.
+func (s *Store) Close() (err error) {
+	s.quitOnce.Do(func() {
+		close(s.quit)
+	})
+
+	timeout := 15 * time.Second
+	done := make(chan struct{})
+	go func() {
+		s.wg.Wait()
+		close(done)
+	}()
+	select {
+	case <-done:
+	case <-time.After(timeout):
+		log.Debug("timeout on waiting chunk store parallel operations to finish", "timeout", timeout)
+	}
+
+	for _, sh := range s.shards {
+		if err := sh.f.Close(); err != nil {
```

Comment:
Keeping the fd open for a long duration without mmap is a disaster in the making. I am not sure if go file flush() does a os level sync(). If it is, then we should flush() at regular intervals to save the contents against crash.

Reply:
Nice observation. Go os.File.Sync() is flushing the content to the disk, I am not sure to which flush() are you referring to.

Flushing on (regular) intervals is what operating system is already doing. I am not sure how this would help agains crash unless we fsync on every write, which is quite costly. I have already tested fsync on every write and it makes fcds much slower, even compared to go-leveldb, as go-leveldb dos not fsync at all.

Mmap brings its own complexity, especially on different operating systems.

Reply:
Thanks for the reply. 

>Go os.File.Sync() is flushing the content to the disk, I am not sure to which flush() are you referring to.

Apologies if i confused you. There are two things... 
1)  flush(), which flushes the application write buffers to OS.
2) fsync(), the OS level sync which absolutely makes sure that buffers has gone to disk. 

The first one is done by golang itself as pointed out by you. The second one is the one i am concerned. 

>Flushing on (regular) intervals is what operating system is already doing

It is usually done by OS disk drivers whenever they feel it is okay.  The fsync() is expensive as pointed out by you. To avoid fsyncing on every commit, DB's usually implement WAL's which fsyncs it on very small regular intervals on the background. so that even if some data is lost it will be of for very small duration . If you fsync on foreground (query path) everytime that will be very expensive.

> even compared to go-leveldb, as go-leveldb dos not fsync at all.

leveldb takes another strategy my doing a larger mmap file than required and written directly using memcopy, then on mmap driver takes care of writing the dirty pages to the disk. 

All i am saying is, one way or other, if we want to evade crash and end up with corrupted files on bootup, We have to implement this otherwise the I am sure we can expect some gibberish files when you switch off power abruptly.

> Mmap brings its own complexity, especially on different operating systems.

Yes. We should not reinvent the wheel. 
As a remedy... we have two options
1) Use already existing DB which takes care of this  ( we can talk about badger if you like)
2) In Forky, do a FIle.Sync() every X seconds ( where we can tolerate X seconds data loss) in a background thread for all files.

I found this intresting read on this topic
https://www.joeshaw.org/dont-defer-close-on-writable-files/

Reply:
Thanks for explanations.

> Use already existing DB which takes care of this ( we can talk about badger if you like)

I have already tried badger and it is slower then go-leveldb. Maybe you can find to do it more efficiently. And still it does not scale with more cpu cores.

> In Forky, do a FIle.Sync() every X seconds ( where we can tolerate X seconds data loss) in a background thread for all files.

I am concerned if this is actually help us, as os is already doing that in some frequency. Also, as you pointed, the solution would be to implement WAL, which in some basic form I already did, with of course, some performance penalties.

As far as I can see go-leveldb is not using mmap.

I see your point very valid, but I think that your suggestions should be tested, even the ones that I already did, to revalidate. We can talk about possibilities, but it would be good to actually test resilience for the weak points that you described.

As this is a very important part of the swarm I am more in favour not to merge this PR until we are clear that it is reliable. Currently, I see that it creates more problems than it brings benefits.

Reply:
I did few changes and did run TestLevelDBForky and TestBadgerSuite test cases 
These are the results.
<img width="1215" alt="Screenshot 2020-03-10 at 1 16 10 AM" src="https://user-images.githubusercontent.com/940575/76251479-0a2e8680-626d-11ea-93f7-e982f4341e29.png">

The  iteration in Badger is somewhat slow... so i commented out the Iteration and did these test cases.. with an assumption that write/read is more important and not iteration.

Reply:
Thanks, but I cannot say anything from the screenshots except to compare timings and conclude that TestBadgerSuite writes are slower. It would be very helpful to actually see the changes and what is TestBadgerSuite doing.

Reply:
Look at the last 2 commits in my fork https://github.com/jmozah/forky/commits/master